### DISCUSSION TITLE
Using same Dataset for multiple tables or list in a report problem

### DISCUSSION THREAD

**Discussion by OnettoHector** [eclipse-birt/birt#1366] [Q&A] [answered]
Hello. Im new to BIRT
I have cretead a scripted dataSource and a scripted dataSet for it.
I have the following code:
In the "Initialize" script of the report:
```
importPackage(Packages.com.datasul.technology.webdesk.dataset);
importPackage(Packages.java.util);

var dataset = null;
var count = 0;
```
In the "beforeOpen" script of the dataset:
```
try{
var factory = DatasetFactory.getInstance(${WKCompany}); 

dataset = factory.getDataset("DSAltaProducto", null, null, null)
}
catch(e){
}
finally{
}
```
In the "fetch" script of the dataset:
```
if (dataset != null) {
if (count < dataset.rowsCount){
	row["codigoProducto"] = dataset.getValue(count, "codigo");
	row["nombreProducto"] = dataset.getValue(count, "nombre");
	count++
	return true; 
}
}
return false;
```
This code seems to work fine. When I drag the dataSource into my design layout it creates a table that displays all of the results just fine.
The problem is, when I try to add another table with the same dataset, nothing is displayed in this second table, its just blank.
Same thing when I create a Data component and bind it to a dataset field, the first one will display fine, and the second one will be blank.
I think im missing an important part about how the dataset operates, so I would appreciate any help.
Thanks in advance

**Comment by hvbtup** [comment, marked as answer]
You have to set count=0 in the `beforeOpen` event (without `var`).
I recommend to use report variables for things like this, just to make it more visible to the next developer that the code needs global variables.

**Reply by OnettoHector** [reply]
Thanks a lot! that fixed it.
May I ask, why is it necessary to reassign the value in the beforeOpen event when I've already assigned it a value in the "Initialize" event?
Im sorry if it's a basic question but I cant seem to find in-depth documentation, do you have a website to point to? I know there's eclipse-birt.github.io/birt-website/docs but the info seems pretty basic
Thanks again!

**Comment by hvbtup** [comment]
Scripted datasets are an advanced topic IMHO. 

Your report is a special case, beause you said that you want to reuse a dataset more than once in the same report.

For those readers who wonder "Why?": There *are* typical use-cases for this scenario (see below).
If your report doesn't fit into these scenarios, please tell us what your use-case is - maybe there is a less complicated solution, then.

Each time you use the data set, you want to iterate over the records.
Thus your record counter has to be reset in the `open` event.
When using a dataset, the sequence of events is `open`, `fetch` (for each record), `close`.

If you don't reset your `count` variable, it is still `after` the end of the records, so at the second time, you'll see no rows.


## Use-cases:

a) You want to show the same data twice in a different form. 
E.g. You start your report with a grouped table with "hide details" checked at some group level, so that the reader can see a summary immediately.
Then, you show the same data again, but now with all the details.
b) You want to re-use the dataset in the same report, but with *different* data (of course the dataset needs data set parameters then). 
c) In *very rare* cases, I use this to combine a list item and a table item based on the same data; I use "hide details" in a group on the list item, then add the table to the group footer, and filter the data so that the outcome is more or less as if I had it all in one table, the difference being that at the outer level I don't have a table, which is useful for MS Word output to avoid nested tables.
